### PR TITLE
Memory and perf fix for Tanh-Sq-Relu and FP32 logits output

### PR DESCRIPTION
**Pull request by xrennvidia**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

- Fuse tanh soft clamp with squared-ReLU in the dense MLP
- Return CE gradient in the logits dtype to avoid autograd's upcast

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Review by fanshiqing — APPROVED** [review]

File: megatron/core/fusions/fused_cross_entropy.py [hunk 1]
```diff
@@ -90,6 +94,9 @@ def forward(ctx, vocab_parallel_logits, target, tp_group):
         """
         Forward implementation for the cross entropy loss.
         """
+        # Record the incoming dtype: calculate_logits_max upcasts to fp32 (a no-op for fp32
+        # logits), and backward must return a gradient of the original dtype.
+        ctx.logits_dtype = vocab_parallel_logits.dtype
```
[the remark this answers is not in the document]
Comment (xrennvidia) [reply megatron/core/fusions/fused_cross_entropy.py:99]: Fixed

**Comment by xrennvidia** [comment]
> Light review — two things, both about coverage/docs rather than the kernels themselves. The math in `clamped_squared_relu` / `clamped_squared_relu_back` checks out against `squared_relu(tanh_soft_clamp(x, s))`, and the new unit tests cover the kernel well (both dtypes, 2D and 3D, the saved-tensor count, and the no-weights-no-clamp assert).
> 
> **1. `use_fused_weighted_squared_relu` docstring is now inaccurate.**
> 
> `megatron/core/transformer/transformer_config.py:548-549` still says the flag only applies to MoE, but the new `mlp.py` branch makes it gate the dense MLP path too:
> 
> ```python
>     use_fused_weighted_squared_relu: bool = False
>     """If True, uses fused weighted squared relu kernel in MoE experts, and the fused
>     tanh-clamp + squared relu kernel in the dense MLP."""
> ```
> 
> (Not an inline suggestion because the file is not in the diff.)
> 
> **2. The new dense-MLP dispatch itself is untested.**
> 
> The unit tests all call `weighted_squared_relu_impl` directly; nothing asserts that `MLP.forward` actually routes to it. All three predicates of the new branch are silent failure modes: flip `self.activation_func == squared_relu` to `self.config.activation_func` (as `experts.py:439-441` deliberately does, with a comment explaining why), drop the flag check, or lose `tanh_clamp_scale is not None`, and every existing test still passes while the fusion quietly stops firing — or fires when it should not.
> 
> Functional CI will not catch it either: no `model_config.yaml` sets `activation_func_tanh_clamp_scale`, and there is no CLI arg for it, so the four recipes that pass `--use-fused-weighted-squared-relu` never reach this branch.
> 
> `tests/unit_tests/transformer/moe/test_grouped_mlp.py:814-834` already has the pattern for this — a monkeypatched-dispatch test parametrized over `(use_fused_weighted_squared_relu, tanh_clamp_scale)` asserting which path is taken. A dense-MLP equivalent in `tests/unit_tests/transformer/test_mlp.py` would pin the branch down without needing a numerics run.

Fixed

**Review by Phlip79 — APPROVED** [review]

**Review by deepakn94 — APPROVED** [review]
